REGRESSION (iOS 18.4.x): Accessing BLOBs stored in IndexedDB leads to Networking process getting terminated

szewai opened the pull request:
#### 74123c260a439c4e69d44300b7ebd2f8a8f39fe7
<pre>
REGRESSION (iOS 18.4.x): Accessing BLOBs stored in IndexedDB leads to Networking process getting terminated
<a href="https://bugs.webkit.org/show_bug.cgi?id=292142">https://bugs.webkit.org/show_bug.cgi?id=292142</a>
<a href="https://rdar.apple.com/150284384">rdar://150284384</a>

Reviewed by Brady Eidson.

When a file-backed Blob is stored in IndexedDB database, we create a new file in the IndexedDB directory to store the
content of the Blob, and make the corresponding IndexedDB record point to the new file instead of the original file.
This ensures websites can have consistent access to the content even when the file is modified or removed. This is
implemented by storing explicit Blob paths (pointing to files in IndexdDB directory) with IndexedDB record. When
IndexedDB client creates file-backed Blob from record, it will prefer using explicit path, and fall back to path in
serialized script value if explicit path does not exist. Normally, IndexedDB server should always send explicit paths to
client. However, in some cases where client retrieves records via IDBIndex, server may not propely include the paths in
result, and this can lead to client tries accessing files that they no longer have access to. To fix this, this patch
makes sure to include Blob paths when server retrieves records with SQLiteIDBCursor.

API test: IndexedDB.GetFileByIndex

* Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
(WebCore::IDBServer::SQLiteIDBCursor::internalFetchNextRecord):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBPersistence.mm:
(-[IndexedDBOpenPanelUIDelegate webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:]):
((IndexedDB, GetFileByIndex)):

Canonical link: <a href="https://commits.webkit.org/297122@main">https://commits.webkit.org/297122@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5966c5ccca17ecb3062cb15da00891b8910aa36f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110549 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20641 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116575 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60816 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112512 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84066 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113497 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24660 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99544 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64507 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24022 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17684 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60370 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94043 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17743 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119365 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37589 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27907 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93032 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37963 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95815 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92855 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23669 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37862 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15604 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33562 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37484 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42955 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37147 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40486 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38855 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5966c5ccca17ecb3062cb15da00891b8910aa36f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47700

szewai committed:
REGRESSION (iOS 18.4.x): Accessing BLOBs stored in IndexedDB leads to Networking process getting terminated
https://bugs.webkit.org/show_bug.cgi?id=292142
rdar://150284384

Reviewed by Brady Eidson.

When a file-backed Blob is stored in IndexedDB database, we create a new file in the IndexedDB directory to store the
content of the Blob, and make the corresponding IndexedDB record point to the new file instead of the original file.
This ensures websites can have consistent access to the content even when the file is modified or removed. This is
implemented by storing explicit Blob paths (pointing to files in IndexdDB directory) with IndexedDB record. When
IndexedDB client creates file-backed Blob from record, it will prefer using explicit path, and fall back to path in
serialized script value if explicit path does not exist. Normally, IndexedDB server should always send explicit paths to
client. However, in some cases where client retrieves records via IDBIndex, server may not propely include the paths in
result, and this can lead to client tries accessing files that they no longer have access to. To fix this, this patch
makes sure to include Blob paths when server retrieves records with SQLiteIDBCursor.

API test: IndexedDB.GetFileByIndex

* Source/WebCore/Modules/indexeddb/server/SQLiteIDBCursor.cpp:
(WebCore::IDBServer::SQLiteIDBCursor::internalFetchNextRecord):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBPersistence.mm:
(-[IndexedDBOpenPanelUIDelegate webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:]):
((IndexedDB, GetFileByIndex)):

Canonical link: https://commits.webkit.org/297122@main

webkit-commit-queue commented:
Committed 297122@main (74123c260a43): <https://commits.webkit.org/297122@main>

Reviewed commits have been landed. Closing PR #47700 and removing active labels.
